# Oakland 311 calls

This notebook is part of Assignment #6.

In [2]:
import pandas as pd
import altair as alt
import requests

## Download data

Download [311 call data](https://data.oaklandca.gov/Infrastructure/Service-requests-received-by-the-Oakland-Call-Cent/quth-gb8e) from the City of Oakland's Open Data Portal. 

In [3]:
# #downloading the Oakland 311 data

# url = 'https://data.oaklandca.gov/api/views/quth-gb8e/rows.csv?accessType=DOWNLOAD'
# r = requests.get(url, allow_redirects=False)

# # write the content of the request into a file called `oakland_311.csv`
# open('oakland_311.csv', 'wb').write(r.content)

# Creating a pandas dataframe

Importing the csv into a pandas dataframe from which I can work on without distorting the original data. I will call the dataframe 'oakland_311_calls_original'


In [4]:
oakland_311_calls_original = pd.read_csv('oakland_311.csv')

# View the data 
Seeking to understand data by looking at the first five rows and the last five rows

In [5]:
oakland_311_calls_original.head()

REQUESTID            DATETIMEINIT       SOURCE  \
0     927708  07/24/2019 07:34:04 AM  SeeClickFix   
1     930512  08/01/2019 01:38:16 PM  SeeClickFix   
2     928898  07/28/2019 09:20:50 AM  SeeClickFix   
3     930760  08/02/2019 10:13:45 AM        Phone   
4     923520  07/11/2019 08:45:50 AM  SeeClickFix   

                            DESCRIPTION REQCATEGORY  \
0           Parking - Abandoned Vehicle      POLICE   
1  Illegal Dumping � mattress/boxspring     ILLDUMP   
2    Human Waste in Public-Right-of-Way       OTHER   
3          City Bldg - Structural Issue   BLDGMAINT   
4         Streets - Potholes/Depression    STREETSW   

                                  REQADDRESS    STATUS  \
0   (37.83332450468423, -122.26187389355277)    CLOSED   
1   (37.80947700189105, -122.29480799834606)    CLOSED   
2  (37.759826001279855, -122.16106799755731)  REFERRED   
3   (37.74458491831174, -122.16914534319535)    CLOSED   
4  (37.780819708391476, -122.22030767672499)    CANCEL   

                             REFERREDTO          DATETIMECLOSED          SRX  \
0                                   NaN  09/30/2019 10:25:11 AM  6052877.252   
1                                   NaN  08/04/2019 10:38:05 AM  6043201.257   
2  ADVANCE CRIME SCENE, RESTORATION EST                     NaN  6081509.580   
3                                   NaN  10/28/2019 11:30:12 AM  6079075.972   
4                                   NaN                     NaN  6064527.852   

           SRY COUNCILDISTRICT BEAT      PROBADDRESS     City State  
0  2130683.728            CCD1  12X      479 45TH ST  Oakland    CA  
1  2122182.593            CCD3  02Y  1117 CHESTER ST  Oakland    CA  
2  2103393.102            CCD7  35X   8601 SENECA ST  Oakland    CA  
3  2097885.948            CCD7  32X  HOLLY MINI PARK  Oakland    CA  
4  2111345.419            CCD5  23X        E 18TH ST  Oakland    CA

In [6]:
oakland_311_calls_original.tail()

REQUESTID            DATETIMEINIT       SOURCE  \
852536     921970  07/07/2019 11:57:45 AM  SeeClickFix   
852537     930717  08/02/2019 08:58:15 AM  SeeClickFix   
852538     928953  07/28/2019 04:50:48 PM  SeeClickFix   
852539     920576  07/01/2019 04:27:51 PM        Phone   
852540     919254  06/27/2019 10:24:31 AM  SeeClickFix   

                                       DESCRIPTION REQCATEGORY  \
852536                       Parking - Enforcement     PARKING   
852537                 Parking - Abandoned Vehicle      POLICE   
852538                 Parking - Abandoned Vehicle      POLICE   
852539  Illegal Dumping - debris, appliances, etc.     ILLDUMP   
852540               Street Light - Outage/Damaged  ELECTRICAL   

                                      REQADDRESS           STATUS REFERREDTO  \
852536  (37.796293496626504, -122.2725785428372)  GONE ON ARRIVAL        NaN   
852537  (37.81960600132651, -122.27702919438389)           CLOSED        NaN   
852538   (37.75789339999198, -122.1766678975261)           CLOSED        NaN   
852539   (37.7697750419514, -122.18324682258614)           CLOSED        NaN   
852540   (37.84077618000984, -122.2688993400894)           CLOSED        NaN   

                DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
852536                     NaN  6049530.979  2117260.646            CCD3  01X   
852537  09/03/2019 09:14:03 AM  6048406.421  2125772.131            CCD3  07X   
852538  12/26/2019 01:18:26 PM  6076988.064  2102769.930            CCD6  30X   
852539  07/01/2019 08:08:55 PM  6075164.293  2107129.719            CCD6  30X   
852540  07/05/2019 02:57:56 PM  6050899.986  2133434.859            CCD1  11X   

                    PROBADDRESS     City State  
852536               308 4TH ST  Oakland    CA  
852537           2825 MARKET ST  Oakland    CA  
852538         7924 plymouth st  Oakland    CA  
852539  66TH AVE & BANCROFT AVE  Oakland    CA  
852540              764 56TH ST  Oakland    CA

# I am also going to inspect the dataframe to understand its datatypes

In [7]:
oakland_311_calls_original.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 852541 entries, 0 to 852540
Data columns (total 16 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   REQUESTID        852541 non-null  int64  
 1   DATETIMEINIT     852541 non-null  object 
 2   SOURCE           852541 non-null  object 
 3   DESCRIPTION      852541 non-null  object 
 4   REQCATEGORY      852460 non-null  object 
 5   REQADDRESS       836654 non-null  object 
 6   STATUS           852540 non-null  object 
 7   REFERREDTO       70168 non-null   object 
 8   DATETIMECLOSED   610875 non-null  object 
 9   SRX              836654 non-null  float64
 10  SRY              836654 non-null  float64
 11  COUNCILDISTRICT  809790 non-null  object 
 12  BEAT             812245 non-null  object 
 13  PROBADDRESS      852064 non-null  object 
 14  City             852541 non-null  object 
 15  State            852541 non-null  object 
dtypes: float64(2), int64(1), object(13)
me

From the code above, I can see that the dataframe has 16 columns identifed by indexes and three dtypes -- int64,object and float64
I can also see that the date has been identified as object instead of datetime64.

# Convert date to datetime64
We can see from the code below that the date ('DATETIMEINIT' and 'DATETIMECLOSED' are actually identified as a str, not dt

In [8]:
oakland_311_calls_original['DATETIMEINIT'].min ()

'01/01/2011 05:51:45 PM'

To correct this, we have to first create a new dataframe from the original so that we always have original data then use pd.to_datetime to convert for both 'DATETIMEINIT' and 'DATETIMECLOSED'

In [9]:
oakland_311 = oakland_311_calls_original.copy()

In [10]:
oakland_311['DATETIMEINIT'] = pd.to_datetime(oakland_311['DATETIMEINIT'])


In [11]:
oakland_311['DATETIMECLOSED'] = pd.to_datetime(oakland_311['DATETIMECLOSED'])

In [12]:
oakland_311.head()

REQUESTID        DATETIMEINIT       SOURCE  \
0     927708 2019-07-24 07:34:04  SeeClickFix   
1     930512 2019-08-01 13:38:16  SeeClickFix   
2     928898 2019-07-28 09:20:50  SeeClickFix   
3     930760 2019-08-02 10:13:45        Phone   
4     923520 2019-07-11 08:45:50  SeeClickFix   

                            DESCRIPTION REQCATEGORY  \
0           Parking - Abandoned Vehicle      POLICE   
1  Illegal Dumping � mattress/boxspring     ILLDUMP   
2    Human Waste in Public-Right-of-Way       OTHER   
3          City Bldg - Structural Issue   BLDGMAINT   
4         Streets - Potholes/Depression    STREETSW   

                                  REQADDRESS    STATUS  \
0   (37.83332450468423, -122.26187389355277)    CLOSED   
1   (37.80947700189105, -122.29480799834606)    CLOSED   
2  (37.759826001279855, -122.16106799755731)  REFERRED   
3   (37.74458491831174, -122.16914534319535)    CLOSED   
4  (37.780819708391476, -122.22030767672499)    CANCEL   

                             REFERREDTO      DATETIMECLOSED          SRX  \
0                                   NaN 2019-09-30 10:25:11  6052877.252   
1                                   NaN 2019-08-04 10:38:05  6043201.257   
2  ADVANCE CRIME SCENE, RESTORATION EST                 NaT  6081509.580   
3                                   NaN 2019-10-28 11:30:12  6079075.972   
4                                   NaN                 NaT  6064527.852   

           SRY COUNCILDISTRICT BEAT      PROBADDRESS     City State  
0  2130683.728            CCD1  12X      479 45TH ST  Oakland    CA  
1  2122182.593            CCD3  02Y  1117 CHESTER ST  Oakland    CA  
2  2103393.102            CCD7  35X   8601 SENECA ST  Oakland    CA  
3  2097885.948            CCD7  32X  HOLLY MINI PARK  Oakland    CA  
4  2111345.419            CCD5  23X        E 18TH ST  Oakland    CA

In [13]:
oakland_311.tail()

REQUESTID        DATETIMEINIT       SOURCE  \
852536     921970 2019-07-07 11:57:45  SeeClickFix   
852537     930717 2019-08-02 08:58:15  SeeClickFix   
852538     928953 2019-07-28 16:50:48  SeeClickFix   
852539     920576 2019-07-01 16:27:51        Phone   
852540     919254 2019-06-27 10:24:31  SeeClickFix   

                                       DESCRIPTION REQCATEGORY  \
852536                       Parking - Enforcement     PARKING   
852537                 Parking - Abandoned Vehicle      POLICE   
852538                 Parking - Abandoned Vehicle      POLICE   
852539  Illegal Dumping - debris, appliances, etc.     ILLDUMP   
852540               Street Light - Outage/Damaged  ELECTRICAL   

                                      REQADDRESS           STATUS REFERREDTO  \
852536  (37.796293496626504, -122.2725785428372)  GONE ON ARRIVAL        NaN   
852537  (37.81960600132651, -122.27702919438389)           CLOSED        NaN   
852538   (37.75789339999198, -122.1766678975261)           CLOSED        NaN   
852539   (37.7697750419514, -122.18324682258614)           CLOSED        NaN   
852540   (37.84077618000984, -122.2688993400894)           CLOSED        NaN   

            DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
852536                 NaT  6049530.979  2117260.646            CCD3  01X   
852537 2019-09-03 09:14:03  6048406.421  2125772.131            CCD3  07X   
852538 2019-12-26 13:18:26  6076988.064  2102769.930            CCD6  30X   
852539 2019-07-01 20:08:55  6075164.293  2107129.719            CCD6  30X   
852540 2019-07-05 14:57:56  6050899.986  2133434.859            CCD1  11X   

                    PROBADDRESS     City State  
852536               308 4TH ST  Oakland    CA  
852537           2825 MARKET ST  Oakland    CA  
852538         7924 plymouth st  Oakland    CA  
852539  66TH AVE & BANCROFT AVE  Oakland    CA  
852540              764 56TH ST  Oakland    CA

Now to confirm whether the dtype for dates changed in the dataframe

In [14]:
oakland_311.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 852541 entries, 0 to 852540
Data columns (total 16 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   REQUESTID        852541 non-null  int64         
 1   DATETIMEINIT     852541 non-null  datetime64[ns]
 2   SOURCE           852541 non-null  object        
 3   DESCRIPTION      852541 non-null  object        
 4   REQCATEGORY      852460 non-null  object        
 5   REQADDRESS       836654 non-null  object        
 6   STATUS           852540 non-null  object        
 7   REFERREDTO       70168 non-null   object        
 8   DATETIMECLOSED   610875 non-null  datetime64[ns]
 9   SRX              836654 non-null  float64       
 10  SRY              836654 non-null  float64       
 11  COUNCILDISTRICT  809790 non-null  object        
 12  BEAT             812245 non-null  object        
 13  PROBADDRESS      852064 non-null  object        
 14  City             852

# Finding the earliest  and latest dates
We will use .min and .max to establish the oldest call and the newest.

In [15]:
oakland_311['DATETIMEINIT'].min()

Timestamp('2009-07-01 08:05:36')

In [16]:
oakland_311['DATETIMECLOSED'].max()

Timestamp('2021-10-29 16:24:38')

From this, we can establish that the first call in the data was made in July, 2009 while the latest was made on Oct 29, 2021.

# Difference between the two dates
To find out how much time has lapsed between the two, we will create a new column in the date called 'DATECLOSED' and populate it with the difference between 'DATETIMECLOSED' and 'DATETIMEINIT'

In [17]:
oakland_311['DATECLOSED'] = oakland_311['DATETIMECLOSED'] - oakland_311['DATETIMEINIT']

In [18]:
oakland_311

REQUESTID        DATETIMEINIT       SOURCE  \
0          927708 2019-07-24 07:34:04  SeeClickFix   
1          930512 2019-08-01 13:38:16  SeeClickFix   
2          928898 2019-07-28 09:20:50  SeeClickFix   
3          930760 2019-08-02 10:13:45        Phone   
4          923520 2019-07-11 08:45:50  SeeClickFix   
...           ...                 ...          ...   
852536     921970 2019-07-07 11:57:45  SeeClickFix   
852537     930717 2019-08-02 08:58:15  SeeClickFix   
852538     928953 2019-07-28 16:50:48  SeeClickFix   
852539     920576 2019-07-01 16:27:51        Phone   
852540     919254 2019-06-27 10:24:31  SeeClickFix   

                                       DESCRIPTION REQCATEGORY  \
0                      Parking - Abandoned Vehicle      POLICE   
1             Illegal Dumping � mattress/boxspring     ILLDUMP   
2               Human Waste in Public-Right-of-Way       OTHER   
3                     City Bldg - Structural Issue   BLDGMAINT   
4                    Streets - Potholes/Depression    STREETSW   
...                                            ...         ...   
852536                       Parking - Enforcement     PARKING   
852537                 Parking - Abandoned Vehicle      POLICE   
852538                 Parking - Abandoned Vehicle      POLICE   
852539  Illegal Dumping - debris, appliances, etc.     ILLDUMP   
852540               Street Light - Outage/Damaged  ELECTRICAL   

                                       REQADDRESS           STATUS  \
0        (37.83332450468423, -122.26187389355277)           CLOSED   
1        (37.80947700189105, -122.29480799834606)           CLOSED   
2       (37.759826001279855, -122.16106799755731)         REFERRED   
3        (37.74458491831174, -122.16914534319535)           CLOSED   
4       (37.780819708391476, -122.22030767672499)           CANCEL   
...                                           ...              ...   
852536   (37.796293496626504, -122.2725785428372)  GONE ON ARRIVAL   
852537   (37.81960600132651, -122.27702919438389)           CLOSED   
852538    (37.75789339999198, -122.1766678975261)           CLOSED   
852539    (37.7697750419514, -122.18324682258614)           CLOSED   
852540    (37.84077618000984, -122.2688993400894)           CLOSED   

                                  REFERREDTO      DATETIMECLOSED          SRX  \
0                                        NaN 2019-09-30 10:25:11  6052877.252   
1                                        NaN 2019-08-04 10:38:05  6043201.257   
2       ADVANCE CRIME SCENE, RESTORATION EST                 NaT  6081509.580   
3                                        NaN 2019-10-28 11:30:12  6079075.972   
4                                        NaN                 NaT  6064527.852   
...                                      ...                 ...          ...   
852536                                   NaN                 NaT  6049530.979   
852537                                   NaN 2019-09-03 09:14:03  6048406.421   
852538                                   NaN 2019-12-26 13:18:26  6076988.064   
852539                                   NaN 2019-07-01 20:08:55  6075164.293   
852540                                   NaN 2019-07-05 14:57:56  6050899.986   

                SRY COUNCILDISTRICT BEAT              PROBADDRESS     City  \
0       2130683.728            CCD1  12X              479 45TH ST  Oakland   
1       2122182.593            CCD3  02Y          1117 CHESTER ST  Oakland   
2       2103393.102            CCD7  35X           8601 SENECA ST  Oakland   
3       2097885.948            CCD7  32X          HOLLY MINI PARK  Oakland   
4       2111345.419            CCD5  23X                E 18TH ST  Oakland   
...             ...             ...  ...                      ...      ...   
852536  2117260.646            CCD3  01X               308 4TH ST  Oakland   
852537  2125772.131            CCD3  07X           2825 MARKET ST  Oakland   
852538  2102769.930            CCD6  30X         7

If we call the first and last five rows, we see that a new column called 'DATECLOSED' has been created with timedelta dtype

# And now to some mean, medium operations...

In [19]:
oakland_311['DATECLOSED'].mean()

Timedelta('49 days 19:45:09.623400859')

In [20]:
oakland_311['DATECLOSED'].median()

Timedelta('4 days 22:03:37')

# ...and the shortest and longest time taken to resolve a case

In [21]:
oakland_311['DATECLOSED'].min()

Timedelta('-2981 days +00:37:47')

#### Not sure why its giving me a negative!!!!

In [22]:
oakland_311['DATECLOSED'].max()

Timedelta('4293 days 21:59:22')

## Let's look at this particular case with -negative close time
Using subsetting

In [23]:
oakland_311[oakland_311['DATECLOSED'] < '0 days']

REQUESTID        DATETIMEINIT          SOURCE  \
123682     432312 2012-11-01 11:24:56  Phone or Email   
229197      23385 2010-09-16 13:59:19  Phone or Email   
229198      23396 2010-09-16 13:59:46  Phone or Email   
229297      24888 2011-04-06 16:33:06  Phone or Email   
229471      25105 2011-05-16 11:55:22  Phone or Email   
...           ...                 ...             ...   
441149     522752 2014-06-10 11:29:36  Phone or Email   
443309     522842 2014-06-10 15:50:45  Phone or Email   
444172     528272 2014-07-10 12:20:22     SeeClickFix   
447612     522805 2014-06-10 14:02:16  Phone or Email   
469154     549531 2014-11-06 14:27:32           Phone   

                                            DESCRIPTION REQCATEGORY  \
123682                                             PG&E       OTHER   
229197             Residential Recycling Service Issues   RECYCLING   
229198             Residential Recycling Service Issues   RECYCLING   
229297  Garbage - Requests Exemption - Weekly or Shared   RECYCLING   
229471             Residential Recycling Service Issues   RECYCLING   
...                                                 ...         ...   
441149             Recycling - Missed Pickup - Backyard   RECYCLING   
443309                  Recycling - Repair/Replace Cart   RECYCLING   
444172                              Hazardous Materials        FIRE   
447612                  Recycling - Repair/Replace Cart   RECYCLING   
469154             Illegal Dumping � mattress/boxspring     ILLDUMP   

                                       REQADDRESS  STATUS  \
123682  (37.790247170707254, -122.23014783691161)  CLOSED   
229197  (30.009899650831787, -141.21875098935422)  CLOSED   
229198  (30.009899650831787, -141.21875098935422)  CLOSED   
229297  (30.009899650831787, -141.21875098935422)  CLOSED   
229471  (30.009899650831787, -141.21875098935422)  CLOSED   
...                                           ...     ...   
441149   (37.81947378040759, -122.19864842876396)  CLOSED   
443309   (37.81389394379093, -122.25288667849271)  CLOSED   
444172  (37.823866701293426, -122.26193144921672)  CLOSED   
447612    (37.82652163613099, -122.2607234958911)  CLOSED   
469154  (37.823866701293426, -122.26193144921672)  CLOSED   

                                  REFERREDTO      DATETIMECLOSED         SRX  \
123682                                 PG&E, 2012-07-10 14:00:14  6061748.00   
229197                                   NaN 2010-09-16 13:59:07        0.00   
229198                                   NaN 2010-09-16 13:59:35        0.00   
229297                                   NaN 2011-04-06 16:33:04        0.00   
229471                                   NaN 2011-05-16 11:55:21        0.00   
...                                      ...                 ...         ...   
441149  CALIFORNIA WASTE SOLUTIONS, EXTERNAL 2014-06-06 16:12:25  6071041.76   
443309  CALIFORNIA WASTE SOLUTIONS, EXTERNAL 2014-06-06 16:12:25  6055339.74   
444172                 FIRE (GRIFFIN, LEROY) 2006-09-08 15:05:19  6052795.80   
447612  CALIFORNIA WASTE SOLUTIONS, EXTERNAL 2014-06-06 16:12:25  6053162.82   
469154                                   NaN 2006-09-08 15:05:19  6052795.80   

               SRY COUNCILDISTRICT BEAT          PROBADDRESS     City State  \
123682  2114830.00             NaN  20X  E 23RD ST & 23RD AV  Oakland    CA   
229197        0.00             NaN  NaN                   ZZ  Oakland    CA   
229198        0.00             NaN  NaN                   ZZ  Oakland    CA   
229297        0.00             NaN  NaN                   ZZ  Oakland    CA   
229471        0.00             NaN  NaN                   ZZ  Oakland    CA   
...            ...             ...  ...                  ...      ...   ...   
441149  2125303.60            CCD4  13Z    2300 MASTLANDS DR  Oakland    CA   
443309  2123561.09            CCD3  14X        302 EUCLID AV  Oakland    CA   
444172  2127240.83            CCD3  08X        MOSSWOOD PARK  Oak

# Which cases took the longest time to process?
## Let's sort the data and find out the top 10

In [32]:
oakland_311.sort_values(by=['DATECLOSED'], ascending=False).head(10)

REQUESTID        DATETIMEINIT          SOURCE  \
264748     323595 2009-12-09 14:30:28  Phone or Email   
273268     357227 2011-01-11 16:11:55         Website   
275539     356540 2011-01-05 21:40:17         Website   
302029     370528 2011-05-18 15:53:54  Phone or Email   
293192     370530 2011-05-18 15:59:39  Phone or Email   
287531     368757 2011-05-02 16:22:17  Phone or Email   
292767     368880 2011-05-03 13:01:49  Phone or Email   
297324     368848 2011-05-03 11:17:04  Phone or Email   
282711     357043 2011-01-10 14:44:07         Website   
266904     356595 2011-01-06 09:25:05         Website   

                                 DESCRIPTION   REQCATEGORY  \
264748           Street Medians (Landscaped)         PARKS   
273268                     Sidewalk - Damage  ROW_STREETSW   
275539  Streets/Sidewalks - Unsafe Condition  ROW_STREETSW   
302029          Park - Landscape Maintenance         PARKS   
293192          Park - Landscape Maintenance         PARKS   
287531  Streets/Sidewalks - Unsafe Condition  ROW_STREETSW   
292767  Streets/Sidewalks - Unsafe Condition  ROW_STREETSW   
297324                     Sidewalk - Damage  ROW_STREETSW   
282711                     Sidewalk - Damage  ROW_STREETSW   
266904                     Sidewalk - Damage  ROW_STREETSW   

                                       REQADDRESS  STATUS REFERREDTO  \
264748   (37.82206394742191, -122.20671960118034)  CLOSED        NaN   
273268   (37.73561875315405, -122.17677744021242)  CLOSED        NaN   
275539    (37.81062199910106, -122.2589020860214)  CLOSED        NaN   
302029   (37.75119515507376, -122.15561532358669)  CLOSED        NaN   
293192   (37.75119515507376, -122.15561532358669)  CLOSED        NaN   
287531  (37.813745661961015, -122.24378665464081)  CLOSED        NaN   
292767    (37.76896035736208, -122.1697033555222)  CLOSED        NaN   
297324  (37.810750150595624, -122.24425986126634)  CLOSED        NaN   
282711    (37.80045375593913, -122.2765999639812)  CLOSED        NaN   
266904   (37.812484999176895, -122.2534134035195)  CLOSED        NaN   

            DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
264748 2021-09-11 12:29:50  6068728.000  2126289.000            CCD4  22X   
273268 2021-08-05 09:29:00  6076811.090  2094661.167            CCD7  31Z   
275539 2021-07-14 09:57:37  6053579.984  2122402.509            CCD3  14X   
302029 2021-09-11 12:08:19  6083030.000  2100223.000            CCD7  35X   
293192 2021-09-11 12:08:47  6083030.000  2100223.000            CCD7  35X   
287531 2021-08-23 15:18:27  6057967.000  2123458.000            CCD2  14Y   
292767 2021-08-03 13:02:52  6079073.000  2106763.000            CCD6  30Y   
297324 2021-07-26 13:07:31  6057810.000  2122370.000            CCD2  14Y   
282711 2021-03-25 13:02:36  6048398.000  2118797.250            CCD3  03Y   
266904 2021-03-18 13:21:37  6055178.000  2123051.000            CCD3  14X   

                      PROBADDRESS     City State         DATECLOSED  
264748  PARK BLVD & MOUNTAIN BLVD  Oakland    CA 4293 days 21:59:22  
273268            10306 PIPPIN ST  Oakland    CA 3858 days 17:17:05  
275539          325 PARK VIEW TER  Oakland    CA 3842 days 12:17:20  
302029            FIRE STATION 26  Oakland    CA 3768 days 20:14:25  
293192            FIRE STATION 26  Oakland    CA 3768 days 20:09:08  
287531            847 WARFIELD AV  Oakland    CA 3765 days 22:56:10  
292767                7521 NEY AV  Oakland    CA 3745 days 00:01:03  
297324          3333 LAKESHORE AV  Oakland    CA 3737 days 01:50:27  
282711      7TH STREET & CLAY ST.  Oakland    CA 3726 days 22:18:29  
266904                377 PALM AV  Oakland    CA 3724 days 03:56:32

# Cleaning the data
Let's first look at the unique identifier and see if all rows have a unique identifier

In [33]:
oakland_311['REQUESTID'].nunique()

852541

### shows there are 852541 unique identifiers. Now let's find out how many rows there are in the dataframe

In [34]:
len(oakland_311)

852541

### Number of unique identifiers matches number of rows

#### So let's check and see if there are duplicated cases using df [expression]

In [36]:
oakland_311[oakland_311.duplicated()]

Empty DataFrame
Columns: [REQUESTID, DATETIMEINIT, SOURCE, DESCRIPTION, REQCATEGORY, REQADDRESS, STATUS, REFERREDTO, DATETIMECLOSED, SRX, SRY, COUNCILDISTRICT, BEAT, PROBADDRESS, City, State, DATECLOSED]
Index: []

Let's narrow this down by using 'REQUESTID'

In [37]:
oakland_311[oakland_311['REQUESTID'].duplicated()]

Empty DataFrame
Columns: [REQUESTID, DATETIMEINIT, SOURCE, DESCRIPTION, REQCATEGORY, REQADDRESS, STATUS, REFERREDTO, DATETIMECLOSED, SRX, SRY, COUNCILDISTRICT, BEAT, PROBADDRESS, City, State, DATECLOSED]
Index: []

## There seems to be no duplicates in the data

# Let's export this data into a csv

To do this, we will use df.to_csv()

In [39]:
oakland_311.to_csv('oakland_311_clean.csv', index=False)